api/task_processor/decorators.py

```diff
@@ -6,107 +6,160 @@
+                    task_identifier=self.task_identifier,
+                    scheduled_for=delay_until or timezone.now(),
+                    priority=self.priority,
+                    queue_size=self.queue_size,
+                    args=args,
+                    kwargs=kwargs,
                 )
+            except TaskQueueFullError as e:
+                logger.warning(e)
                 return
 
-            if settings.TASK_RUN_METHOD == TaskRunMethod.SYNCHRONOUSLY:
-                _validate_inputs(*args, **kwargs)
-                f(*args, **kwargs)
-            elif settings.TASK_RUN_METHOD == TaskRunMethod.SEPARATE_THREAD:
-                logger.debug("Running task '%s' in separate thread", task_identifier)
-                run_in_thread(args=args, kwargs=kwargs)
-            else:
-                logger.debug("Creating task for function '%s'...", task_identifier)
-                try:
-                    task = Task.create(
-                        task_identifier=task_identifier,
-                        scheduled_for=delay_until or timezone.now(),
-                        priority=priority,
-                        queue_size=queue_size,
-                        args=args,
-                        kwargs=kwargs,
-                    )
-                except TaskQueueFullError as e:
-                    logger.warning(e)
-                    return
-
-                task.save()
-                return task
-
-        def run_in_thread(*, args: typing.Tuple = (), kwargs: typing.Dict = None):
-            logger.info("Running function %s in unmanaged thread.", f.__name__)
-            _validate_inputs(*args, **kwargs)
-            Thread(target=f, args=args, kwargs=kwargs, daemon=True).start()
-
-        def _wrapper(*args, **kwargs):
-            """
-            Execute the function after validating the arguments. Ensures that, in unit testing,
-            the arguments are validated to prevent issues with serialization in an environment
-            that utilises the task processor.
-            """
-            _validate_inputs(*args, **kwargs)
-            return f(*args, **kwargs)
-
-        _wrapper.delay = delay
-        _wrapper.run_in_thread = run_in_thread
-        _wrapper.task_identifier = task_identifier
-
-        # patch the original unwrapped function onto the wrapped version for testing
-        _wrapper.unwrapped = f
-
-        return _wrapper
+            task.save()
+            return task
+
+    def run_in_thread(
+        self,
+        *,
+        args: tuple[typing.Any] = (),
+        kwargs: dict[str, typing.Any] | None = None,
+    ) -> None:
+        _validate_inputs(*args, **kwargs)
+        thread = Thread(target=self.unwrapped, args=args, kwargs=kwargs, daemon=True)
+
+        def _start() -> None:
+            logger.info(
+                "Running function %s in unmanaged thread.", self.unwrapped.__name__
+            )
+            thread.start()
+
+        if self.transaction_on_commit:
```

Comment:
Maybe add a comment here (or in the docstring below) to explain why we have this option?

Reply:
Done.